spark-jobserver/spark-jobserver
# feat(integration-tests): Add subproject for integration tests

I wanted to share this integration test suite that I developed over time, because I thought this would be interesting for the community as well.

**Motivation**: Unit testing is great, but what I sometimes miss when developing on job server is a possibility to test the whole deployment stack (including spark, databases, VMs, network, HA, ...), which at least in my setup is sometimes hard to containerize.

**Goal**: Ultimately I wanted something like a job server black box test, meaning it should check if a deployment fulfils the "job server functionality contract" (i.e.: Does the API work as intended, does HA failover work without downtime, does everything come up after a restart, ...) externally. The only two access points would be WebAPI requests and starting/stopping a jobserver.

**Realization**: 
* `job-server-integration-tests` is a separate project which is built independently to one fat jar and can then be run in any environment.
* I added a small documentation on how to run (and configure) the test suite
* The test suite is configurable: You can decide which tests to run, which jobserver addresses to test against, if https is used and which kind of deployment you are running.
* The test suite contains the following tests: `BasicApiTests` test all of the available WebApi, `CornerCasesTests` check more "complex" api logic, `TwoJobserversTests` check if an active-active HA deployment works and `HAFailoverTest` check if an HA jobserver is still working after taking one jobserver down.
* I abstracted away a `DeploymentController` interface, meaning some (HA failover) tests might need to start/stop a jobserver instance and you can theoretically do that with any kind of deployment: Locally, docker, ... . Anyone who is interested in running such kind of tests can add a small class implementing this interface specific to his deployment and is ready to go.

**Tests:**
This is the complete (current) list of tests executed when running all four test classes:
```
BasicApiTests:
/binaries
- POST /binaries/<app> should upload a binary
- GET /binaries should list all available binaries
- GET /binaries/<app> should retrieve a specific binary
- POST /binaries/<app> should overwrite a binary with a new version
- DELETE /binaries/<app> should delete all binary versions under this name
  Error scenarios
  - GET /binaries/<app> should fail if the binary does not exist
  - DELETE /binaries<app> should fail if the binary does not exist
/contexts
- POST /contexts/<contextName> should create a new context
- GET /contexts should list all contexts
- GET /contexts/<contextName> should retrieve infos about a context
- DELETE /contexts/<contextName> should delete a context
- GET /contexts should not list deleted contexts
  Error scenarios
  - POST /contexts/<contextName> should fail if the context name already exists
  - DELETE /contexts/<contextName> should fail if there is no such context
/jobs
- (preparation) uploading job binaries for job testing should not fail
  adHoc jobs
  - POST /jobs?cp=..&mainClass=.. should start a job
  - the termination of the job should also terminate the adHoc context
  batch jobs
  - POST /jobs?context=<context>&cp=..&mainClass=., should start a job in an existing (batch) context
  - the termination of the job should not terminate the context
  streaming jobs
  - POST /jobs?context=<streamingContext> should start a job in an existing (streaming) context
  - DELETE /jobs/<jobId> should stop a streaming job
  - the termination of the streaming job should not terminate the streaming context
- POST /jobs?cp=..&mainClass=.. should start a job from 2 binaries
- POST /jobs should start a job by taking cp and mainClass values from config
- GET /jobs should list all jobs
- GET /jobs/<id> should show job information
- GET /jobs/<id>/config should return the job config
  legacy API (appName + classPath)
    adHoc jobs
    - POST /jobs?appName=..&classPath=.. should start a job in adHoc context
    - the termination of the job should also terminate the adHoc context
    batch jobs
    - POST /jobs?context=<context> should start a job in an existing (batch) context
    - the termination of the job should not terminate the context
    streaming jobs
    - POST /jobs?context=<streamingContext> should start a job in an existing (streaming) context
    - DELETE /jobs/<jobId> should stop a streaming job
    - the termination of the streaming job should not terminate the streaming context
  Error scenarios
  - POST /jobs should fail if there is no such app
  - POST /jobs should fail if there is no such context
  - GET /jobs/<id> should return an error if there is no such id
  - GET /jobs/<id>/config should return an error if there is no such id
  - DELETE /jobs/<id> should return an error if there is no such id
  - POST /jobs should fail if configured improperly (cp is missing)
  - POST /jobs should fail if configured improperly (mainClass is missing)
  - POST /jobs should fail if configured improperly (cp and classPath)
  - POST /jobs should fail if configured improperly (appName and mainClass)
  - POST /jobs should return BadRequest if URI has invalid protocol
CornerCasesTests:
- DELETE /binaries should not delete binaries with running jobs
TwoJobserverTests:
Synchronize contexts across jobservers
- POST /context on sjs1 should be visible on sjs2
- POST /context on sjs2 should be visible on sjs1
- DELETE /context on sjs1 should be visible on sjs2
- DELETE /context on sjs2 should be visible on sjs1
HAFailoverTest:
Sustain one jobserver failure
- Stopping Jobserver 1 should succeed
- GET /binaries should still work on Jobserver 2
- waiting for 30s until the cluster has fully recovered...
- POST /contexts should still work on Jobserver 2
- GET /contexts should still work on Jobserver 2
- DELETE /contexts should still work on Jobserver 2
- Restart of Jobserver 1 should succeed
- GET /binaries should work again on Jobserver 1
- POST /contexts should work again on Jobserver 1
- GET /contexts should work again on Jobserver 1
- DELETE /contexts should work again on Jobserver 1
```


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/spark-jobserver/spark-jobserver/1285)
